Library/Homebrew/formula.rb

```diff
@@ -1625,6 +1625,49 @@ def extract_macho_slice_from(file, arch = Hardware::CPU.arch)
   end
   private :extract_macho_slice_from
 
+  # Generate shell completions for a formula for bash, zsh, and fish, using the formula's binary.
+  #
+  # @param base_name [String] the base name of the generated completion script. Defaults to the formula name.
+  # @param shells [Array<Symbol>] the shells to generate completion scripts for. Defaults to `[:bash, :zsh, :fish]`.
+  # @param binary [Pathname] the binary to use for generating the completion scripts. Defaults to the binary with the
+  # name of the formula.
+  # @param cmd [String] the command to pass to the `binary`. Defaults to 'completion'.
+  # @param shell_as_flag [Boolean] specify if `shells` should each be passed as flags to the `binary`.
+  # Defaults to `false`.
+  sig {
+    params(base_name: String, shells: T::Array[Symbol], binary: Pathname, cmd: String,
+           shell_prefix: T.nilable(T.any(Symbol, String))).void
+  }
+  def generate_completions_from_executable(base_name: name,
+                                           shells: [:bash, :zsh, :fish],
+                                           executable: bin/base_name,
+                                           cmd: "completion",
+                                           shell_parameter_format: nil)
+    completion_script_path_map = {
+      bash: bash_completion/base_name,
+      zsh:  zsh_completion/"_#{base_name}",
+      fish: fish_completion/"#{base_name}.fish",
+    }
+
+    shells.each do |shell|
+      script_path = completion_script_path_map[shell]
+      shell_parameter = if shell_parameter_format.nil?
+        shell.to_s
+      elsif shell_parameter_format == :flag
+        "--#{shell}"
+      elsif shell_parameter_format == :arg
+        "--shell=#{shell}"
+      elsif shell_parameter_format == :none
+        nil
+      else
+        "#{shell_parameter_format}#{shell}"
+      end
+
+      script_path.dirname.mkpath
```

Comment:
I'm curious about why this is needed since many formulae seem to be able to do something like `script_path.write` directly.

Reply:
I was wondering that too, but I am not all too familiar with all the `brew` internals.
Without this, something like this errors out, though.
```
❯ b reinstall k3sup -s
==> Cloning https://github.com/alexellis/k3sup.git
Updating /Users/Max/Library/Caches/Homebrew/k3sup--git
==> Checking out tag 0.12.0
HEAD is now at c59d67b Update Go crypo libraries and migrate to Go 1.18
HEAD is now at c59d67b Update Go crypo libraries and migrate to Go 1.18
==> Reinstalling k3sup
==> go build -ldflags=-s -w -X github.com/alexellis/k3sup/cmd.Version=0.12.0 -X github.com/alexellis/k3sup/cmd.GitCommit=c59d67b
Error: An exception occurred within a child process:
  Errno::ENOENT: No such file or directory @ rb_sysopen - /usr/local/Cellar/k3sup/0.12.0/etc/bash_completion.d/k3sup
```